### PR TITLE
rix(repeat): fix keyed issues & contextual property optimization

### PR DESCRIPTION
**Pull request by fkleuver**
# Pull Request

## 📖 Description

- Adds an optimization for contextual properties
- Fixes contextual properties for duplicate items (including duplicate primitives)
- Significant simplification of update logic by using scope objects as the keys in non-keyed mode
- Memory usage improvements

### 🎫 Issues
https://github.com/aurelia/aurelia/issues/2037

### CODE REVIEW

File: packages/runtime-html/src/binding/scope.ts [hunk 1]
```diff
@@ -102,11 +102,11 @@ export class Scope {
     return new Scope(null, bc as IBindingContext, oc ?? new OverrideContext(), isBoundary ?? false);
   }
 
-  public static fromParent(ps: Scope | null, bc: object): Scope {
+  public static fromParent(ps: Scope | null, bc: object, oc: IOverrideContext = new OverrideContext()): Scope {
```
Comment (bigopon) [inline packages/runtime-html/src/binding/scope.ts:105]: 👍 there' been a few times I added this but removed in the end

File: packages/runtime-html/src/resources/template-controllers/repeat.ts [hunk 2]
```diff
@@ -70,6 +69,8 @@ export class Repeat<C extends Collection = unknown[]> implements ICustomAttribut
 
   public views: ISyntheticView[] = [];
   private _oldViews: ISyntheticView[] = [];
+  private _scopes: Scope[] = [];
+  private _oldScopes: Scope[] = [];
```
Comment (bigopon) [inline packages/runtime-html/src/resources/template-controllers/repeat.ts:73]: should have internal for these

File: packages/runtime-html/src/resources/template-controllers/repeat.ts [hunk 3]
```diff
@@ -723,16 +630,31 @@ interface IRepeatOverrideContext extends IOverrideContext {
...
+  public get $middle(): boolean {
+    return !this.$first && !this.$last;
+  }
+  public get $last(): boolean {
+    return this.$index === this.$length - 1;
+  }
+
+  public constructor(
+    public readonly $index: number = 0,
+    public readonly $length: number = 1,
```
Comment (bigopon) [inline packages/runtime-html/src/resources/template-controllers/repeat.ts:652]: likely a weird default but these won't be use so maybe it's fine

File: packages/runtime-html/src/resources/template-controllers/repeat.ts [hunk 4]
```diff
@@ -588,31 +509,17 @@ export class Repeat<C extends Collection = unknown[]> implements ICustomAttribut
...
-        }
-        setContextualProperties(view.scope.overrideContext as IRepeatOverrideContext, i, newLen);
+        setContextualProperties(view.scope.overrideContext as RepeatOverrideContext, i, newLen);
         view.nodes.insertBefore(view.location!);
       } else {
-        if (_hasDestructuredLocal) {
-          astAssign(dec, view.scope, _forOfBinding, _normalizedItems![i]);
-        } else {
-          view.scope.bindingContext[local] = _normalizedItems![i];
-        }
```
Comment (bigopon) [inline packages/runtime-html/src/resources/template-controllers/repeat.ts:612]: this seems like a good cleanup, items won't be retrieved/checked in so many places

File: packages/runtime-html/src/resources/template-controllers/repeat.ts [hunk 5]
```diff
@@ -588,31 +509,17 @@ export class Repeat<C extends Collection = unknown[]> implements ICustomAttribut
       view.nodes.link(next?.nodes ?? _location);
 
       if (indexMap[i] === -2) {
-        viewScope = getScope(_scopeMap, _normalizedItems![i] as IIndexable, forOf, parentScope, _forOfBinding, local, _hasDestructuredLocal);
```
Comment (bigopon) [inline packages/runtime-html/src/resources/template-controllers/repeat.ts:591]: 👍

**Review by bigopon — APPROVED** [review]
Nice work @fkleuver , a nice fix + a good cleanup 👍

**Comment by ekzobrain** [comment]
Does this change fix a sorting issue when the same objects are reordered in an array? In v21 contextual properties brake after that.
Example:
```js
const a = {};
const b = {};
const c = {};
const d = {};

let arr = [a, b, c, d];

// when rendering this list objects get $index'es a=0, b=1, c=2, d=3
// then update update list using the same object in another order:

arr = [b, c, d, a]

// when rendering updated list $index'es are broken: b=0, c=1, d=2, a=0 (removed and reinserted at other position items keep their previous index)
```

**Comment by ekzobrain** [comment]
And also this race condition issue? https://discord.com/channels/448698263508615178/448699089513611266/1274338349297041489

**Comment by bigopon** [comment]
> And also this race condition issue? https://discord.com/channels/448698263508615178/448699089513611266/1274338349297041489

@ekzobrain list mutation while rendering will be a separate issue. This issue is about correctly tracking items movement when there's duplicate primitives

**Comment by ekzobrain** [comment]
> > And also this race condition issue? https://discord.com/channels/448698263508615178/448699089513611266/1274338349297041489
> 
> 
> 
> @ekzobrain list mutation while rendering will be a separate issue. This issue is about correctly tracking items movement when there's duplicate primitives

I asket just in case it was touched alongside )

**Comment by fkleuver** [comment]
> Does this change fix a sorting issue when the same objects are reordered in an array? In v21 contextual properties break after that. Example:
> 
> ```js
> const a = {};
> const b = {};
> const c = {};
> const d = {};
> 
> let arr = [a, b, c, d];
> 
> // when rendering this list objects get $index'es a=0, b=1, c=2, d=3
> // then update update list using the same object in another order:
> 
> arr = [b, c, d, a]
> 
> // when rendering updated list $index'es are broken: b=0, c=1, d=2, a=0 (removed and reinserted at other position items keep their previous index)
> ```

Yes, there is logic in place that ensures the index is updated in all circumstances, including the scenario you mentioned

**Pull request merged**